### PR TITLE
CV3-61-story(add-bouquet): Implement addBouquet function

### PR DESCRIPTION
**Pull request by kodek-sleuth**
## Description ##
This pull request implements functionality that enables addition of a bouquet.

## How can This be tested ##
- Clone the repo: `git clone https://github.com/andela/mrm_push.git`
- Setup project according to `Readme.md`
- checkout to branch `story/CV3-61-implement-addbouquet-functionality`
- Get credentials from Google [visit](https://developers.google.com/calendar/quickstart/python)
- Download `credentials.json` file and insert it into root folder 
- to add bouquet, hit the endpoint `/v2/bouquets` with a post method and supply the following data
```
auth_uri: 
bouquet_name: ,
client_id: 
client_secret: 
redirect_uris: 
refresh_url
should_refresh
token_uri:
```

## Type of change ##
Please select the relevant option
- [ ] Bugfix(a non-breaking change which fixes an issue)
- [x] New feature(a non-breaking change which adds functionality)
- [ ] Breaking change(fix of a feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

## Checklist: ##
- [x] My code follows the style guidelines of this project
- [x] I have linted my code prior to submission
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing tests pass locally with my changes

## JIRA ##
[CV3-61](https://andela-apprenticeship.atlassian.net/browse/CV3-61?atlOrigin=eyJpIjoiMGU5NzQ5NjY2YWI5NDJlZDhhMmNhYmZjYTU1ZDVlNzciLCJwIjoiaiJ9)

### CODE REVIEW

File: api/v2/models/bouquets/bouquets_model.py [hunk 1]
```diff
@@ -40,8 +38,11 @@ def receive_notification_from_google(self):
 
         # TODO: add functionality to receive calendar notifications for purposes of logging them on the relay log table
 
-    def add_bouquet(self, bouquet_id):
+    @staticmethod
+    def add_bouquet(**kwargs):
         """Method for adding a bouquet"""
+        x = Bouquets(**kwargs)
```
Comment (MCFrank16) [inline api/v2/models/bouquets/bouquets_model.py]: this variable X isn't descriptive at all. if not mistaken though

File: api/v2/models/bouquets/bouquets_model.py [hunk 2]
```diff
@@ -40,8 +38,11 @@ def receive_notification_from_google(self):
 
         # TODO: add functionality to receive calendar notifications for purposes of logging them on the relay log table
```
Comment (MCFrank16) [inline api/v2/models/bouquets/bouquets_model.py:43]: is this something that you will have to work on later on?

File: api/v2/models/bouquets/bouquets_schema.py [hunk 3]
```diff
@@ -5,5 +5,4 @@
 class BouquetsSchema(ma.Schema):
     class Meta:
         model = BouquetsModel
-        fields = ("id", "bouquet_id", "api_key1",  "api_key2"
-                  "auth_credentials", "bouquet_name", "should_refresh")
+        fields = ("id", "bouquet_id", "api_key1",  "api_key2", "auth_credentials", "bouquet_name", "should_refresh")
```
Comment (MCFrank16) [inline api/v2/models/bouquets/bouquets_schema.py]: this should_refresh field name indicates what exactly?? what is its purpose, I am totally confused @kodek-sleuth

**Review by MCFrank16 — APPROVED** [review]

File: api/v2/models/bouquets/bouquets_schema.py (hunk 3 above)
Comment (kodek-sleuth) [reply api/v2/models/bouquets/bouquets_schema.py]: Imagine if in the future we were to supply a job(celery) to this function to always refresh and get channels from the refresh url, this would be the basis on which we would base from.

File: api/v2/models/bouquets/bouquets_model.py (hunk 2 above)
Comment (kodek-sleuth) [reply api/v2/models/bouquets/bouquets_model.py:43]: Yes, thats functionality that will be added later on

File: api/v2/models/bouquets/bouquets_model.py (hunk 1 above)
Comment (kodek-sleuth) [reply api/v2/models/bouquets/bouquets_model.py]: Right, let me change this asap

**Review by Emile-Nsengimana — COMMENTED** [review]
@kodek-sleuth  can you change the api route in the `How can This be tested` section of the `README` to `/v2/bouquets` instead of `v2/boquets`(it is misleading)

Add open `[` on your commit message footer

**Comment by Emile-Nsengimana** [comment]
I can't seem to add a bouquet even after providing all required field
- request data: 
```
{
        "bouquet_name": "Andela bouquet",
	"api_key1": "jnjnjcndjncjndjcnjdncjnjdcnjdncjndcjndjcnjdnc",
	"api_key2": "jnjnjcndjncjndjcnjdncjnjdcnjdncjndcjndjcnjdnc",
	"should_refresh": true,
	"auth_credentials": "jnjnjcndjncjndjcnjdncjnjdcnjdncjndcjndjcnjdnc",
	"refresh_url":"http://localhost:5000/v2/bouquets"
}
```
- response:

    The response is not helping to identify what's wrong
     ```
    {
       "error": "failed to add bouquet"
    }
   ```

**Review by Emile-Nsengimana — APPROVED** [review]

**Review by Emile-Nsengimana — APPROVED** [review]

File: api/v2/helpers/bouquets/bouquets_helper.py [hunk 4]
```diff
@@ -9,3 +9,15 @@ def query_all_bouquets():
         bouquets = BouquetsSchema(many=True).dump(all_bouquets)
 
     return bouquets
+
+
+def create_bouquet(request_data):
```
Comment (joshuaocero) [inline api/v2/helpers/bouquets/bouquets_helper.py]: This function is too brief, specific and concerns mostly the model. It does not deserve to be a helper function.

File: api/v2/helpers/bouquets/bouquets_helper.py [hunk 5]
```diff
@@ -9,3 +9,15 @@ def query_all_bouquets():
...
+    """Function to add bouquet to database
+        :param request_data Required
+            Post data sent to server
+    """
+    BouquetsModel.add_bouquet(bouquet_name=request_data['bouquet_name'], refresh_url=request_data['refresh_url'],
```
Comment (joshuaocero) [inline api/v2/helpers/bouquets/bouquets_helper.py]: Here is the complexity I was talking about. It is not enough to add the record to the database. You need to test the credentials and API Keys passed to make sure they are good enough to make a connection to the Google Calendar API. We should not be storing invalid credentials. This will be the hard nut to crack.

File: api/v2/utilities/validators.py [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+def validate_bouquet_adding(func):
+    """Decorator to validate the addition of a bouquet"""
+
+    @wraps(func)
+    def wrapper():
+        if func.__name__ == 'post':
+            request_data = request.get_json()
+
+            def check_numeric(_value):
+                _re = bool(re.fullmatch(r'^[a-zA-Z]*\s?[a-zA-Z]*\s?[a-zA-Z]*', str(value)))
```
Comment (joshuaocero) [inline api/v2/utilities/validators.py]: Python has a built-in `isnumeric` function that can save you the regular expression.

File: api/v2/utilities/validators.py [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+                return _re
+
+            def check_url(url):
+                _re = bool(re.fullmatch(
```
Comment (joshuaocero) [inline api/v2/utilities/validators.py]: There are also URL checking libraries that can be helpful here like `urlparse` which is inbuilt in Python and other external libraries.

File: api/v2/utilities/validators.py [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+                    r'(https?:\/\/(?:www\.|(?!www))[a-zA-Z0-9][a-zA-Z0-9-]+[a-zA-Z0-9]\.[^\s]{2,}|'
+                    r'www\.[a-zA-Z0-9][a-zA-Z0-9-]+[a-zA-Z0-9]\.[^\s]{2,}|'
+                    r'https?:\/\/(?:www\.|(?!www))[a-zA-Z0-9]+\.[^\s]{2,}|'
+                    r'www\.[a-zA-Z0-9]+\.[^\s]{2,})', url))
+                return _re
+
+            try:
```
Comment (joshuaocero) [inline api/v2/utilities/validators.py:16]: You should consider flask restful's `reqparse` to handle the key validation you are doing here. It is a lot more cleaner. Check out this link -> `https://flask-restful.readthedocs.io/en/0.3.5/reqparse.html

**Review by joshuaocero — CHANGES_REQUESTED** [review]
Great job so far @kodek-sleuth. Please see my feedback and please take note of the complexity I mentioned.

File: alembic/versions/12beb79ff598_add_bouquets_table.py [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+# revision identifiers, used by Alembic.
+revision = '12beb79ff598'
+down_revision = 'd20415dbe205'
+branch_labels = None
+depends_on = None
+
+
+def upgrade():
+    # ### commands auto generated by Alembic - please adjust! ###
+    op.create_table('bouquets',
```
Comment (joshuaocero) [inline alembic/versions/12beb79ff598_add_bouquets_table.py]: This table already exists, so does the `channels` table.

**Review by joshuaocero — CHANGES_REQUESTED** [review]
@kodek-sleuth please correct your alembic file. The tables you are creating in your upgrade already exist. They were already created by the first two migrations.

**Review by joshuaocero — CHANGES_REQUESTED** [review]
@kodek-sleuth you should never edit previous alembic files that have been applied. That is just bad practice. The correct way is to create alembic files to make modifications.

File: alembic/versions/8efbde3ecaeb_alter_bouquets_table.py [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+branch_labels = None
+depends_on = None
+
+def upgrade():
+    # ### commands auto generated by Alembic - please adjust! ###
+    with op.batch_alter_table('bouquets') as batch_op:
+        batch_op.alter_column('auth_credentials', new_column_name='auth_uri')
+        batch_op.alter_column('api_key1', new_column_name='token_uri')
+        batch_op.alter_column('api_key2', new_column_name='client_id')
+    # op.add_column('bouquets', sa.Column('auth_uri', sa.String(), nullable=True))
```
Comment (joshuaocero) [inline alembic/versions/8efbde3ecaeb_alter_bouquets_table.py]: Please remove this comment

File: alembic/versions/c5d7efe7e52c_initial.py [hunk 11]
```diff
@@ -1,10 +1,11 @@
 """initial
 
 Revision ID: c5d7efe7e52c
-Revises: 
+Revises:
```
Comment (joshuaocero) [inline alembic/versions/c5d7efe7e52c_initial.py]: Undo this change. We must never modify previous migration files.

**Review by joshuaocero — CHANGES_REQUESTED** [review]
@kodek-sleuth please see my requests changes, thanks.

File: api/v2/models/bouquets/bouquets_model.py [hunk 12]
```diff
@@ -1,29 +1,31 @@
-from sqlalchemy import (Column, String, Integer, Boolean, Enum)
+from sqlalchemy import *
```
Comment (joshuaocero) [inline api/v2/models/bouquets/bouquets_model.py]: Why do you introduce wildcard an import here? Please read import best practice here: https://www.python.org/dev/peps/pep-0008/#imports

File: api/v2/models/bouquets/bouquets_schema.py (hunk 3 above)
Comment (joshuaocero) [inline api/v2/models/bouquets/bouquets_schema.py]: This change does nothing but breaks PEP 8 recommendation on maximum line length. https://www.python.org/dev/peps/pep-0008/#maximum-line-length

File: test_push.py [hunk 13]
```diff
@@ -7,7 +7,7 @@
 from app import create_app
 from api.v1.service.push_notification import PushNotification
 
-class api_test_case(unittest.TestCase):
+class ApiTestCase(unittest.TestCase):
```
Comment (joshuaocero) [inline test_push.py:12]: Great catch!

File: tests/base.py [hunk 14]
```diff
@@ -27,30 +27,102 @@ def setUp(self):
         self.app_test = app.test_client()
         with app.app_context():
             Base.metadata.create_all(bind=engine)
-
             command.stamp(self.alembic_configuration, 'head')
-            command.downgrade(self.alembic_configuration, '-1')
+            command.downgrade(self.alembic_configuration, 'base')
```
Comment (joshuaocero) [inline tests/base.py]: Why would you introduce this change to downgrade all the way to the base? It is not right and is not in the scope of your task.

File: tests/base.py [hunk 15]
```diff
@@ -27,30 +27,102 @@ def setUp(self):
...
             command.upgrade(self.alembic_configuration, 'head')
 
+            self.bouquet_201 = {
```
Comment (joshuaocero) [inline tests/base.py:36]: How come you are getting authentication details from the environment variables yet we are now storing them in the db?

File: tests/test_channels/test_channels.py [hunk 16]
```diff
@@ -1,8 +1,6 @@
 from tests.base import BaseTestCase
 
-
```
Comment (joshuaocero) [inline tests/test_channels/test_channels.py]: Why do you remove this line? Please see PEP * recommendations on blank lines before classes: https://www.python.org/dev/peps/pep-0008/#blank-lines

File: tests/test_bouquets/test_bouquets.py [hunk 17]
```diff
@@ -1,8 +1,51 @@
+import json
+from unittest.mock import patch
 from tests.base import BaseTestCase
 
+def calendar():
```
Comment (joshuaocero) [inline tests/test_bouquets/test_bouquets.py]: Having this function here is not neat and is duplication. You should rather call the main function and have it behave differently for tests. This will also ensure you have good coverage of this part of the code.

File: api/v2/controllers/bouquets/bouquets_controller.py [hunk 18]
```diff
@@ -0,0 +1,34 @@
...
+        data = request.get_json()
+        redirect_uris = ''
+        if 'should_refresh' not in data.keys():
+            data['should_refresh'] = False
+        if 'refresh_url' not in data.keys():
+            data['refresh_url'] = 'http://www.google.com'
+        for i in data['redirect_uris']:
+            redirect_uris += f'{i} '
+        data['redirect_uris'] = redirect_uris
+        result = calendar(data)
```
Comment (joshuaocero) [inline api/v2/controllers/bouquets/bouquets_controller.py]: This function `calendar` is being called in both the validator and in the `post` method. Aren't you doing double work?

File: api/v2/controllers/bouquets/bouquets_controller.py [hunk 19]
```diff
@@ -0,0 +1,34 @@
...
+        BouquetsModel.add_bouquet(bouquet_name=data['bouquet_name'], client_id=data['client_id'],
```
Comment (joshuaocero) [inline api/v2/controllers/bouquets/bouquets_controller.py:29]: What happens if the credentials do not work? We should not store the bouquet if the result is negative.

File: alembic/versions/8efbde3ecaeb_alter_bouquets_table.py [hunk 20]
```diff
@@ -0,0 +1,47 @@
+"""alter bouquets table
+
+Revision ID: 8efbde3ecaeb
+Revises: ffc544c2bbe6
```
Comment (joshuaocero) [inline alembic/versions/8efbde3ecaeb_alter_bouquets_table.py]: Kindly update this comment

File: api/v2/helpers/credentials.py [hunk 21]
```diff
@@ -0,0 +1,51 @@
...
+from google_auth_oauthlib.flow import InstalledAppFlow, Flow
+from google.auth.transport.requests import Request
+
+from api.v2.models.bouquets.bouquets_model import Bouquets
+
+# If modifying these scopes, delete the file token.pickle.
+SCOPES = ['https://www.googleapis.com/auth/calendar.readonly']
+
+
+def calendar(data):
```
Comment (joshuaocero) [inline api/v2/helpers/credentials.py]: This is a bad function name. Function names should have descriptive verbs. This looks like a variable name.

**Review by joshuaocero — CHANGES_REQUESTED** [review]
@kodek-sleuth I've found quite a lot of issues on this PR. Please look into them and I implore to go through the PEP 8 doc to learn more on Python best practice.

File: api/v2/controllers/bouquets/bouquets_controller.py (hunk 19 above)
Comment (kodek-sleuth) [reply api/v2/controllers/bouquets/bouquets_controller.py:29]: The server sends a fallback message that says the operation failed

File: tests/base.py (hunk 15 above)
Comment (kodek-sleuth) [reply tests/base.py:36]: For testing purposes

File: api/v2/controllers/bouquets/bouquets_controller.py (hunk 19 above)
Comment (joshuaocero) [reply api/v2/controllers/bouquets/bouquets_controller.py:29]: Alright, makes sense.

**Review by joshuaocero — APPROVED** [review]
